code/part-two/client/source/MojiListItem.jsx

```diff
@@ -9,47 +9,18 @@ export class MojiListItem extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
-      isLoaded: false,
       isSire: false,
       mojiView: null
     };
   }
 
-  componentDidMount() {
-    Promise.resolve(this.props.address)
-      .then(address => {
-        // if it's an address, then get the details
-        if (!!address) {
-          return getMoji(this.props.address);
-        }
-        // otherwise, the details were already passed in
-        return this.props.moji;
-      })
-      .then(moji => {
-        this.setState({
-          mojiView: parseDna(moji.dna).view
-        });
+  async componentDidMount() {
+    let moji = this.props.moji || await getMoji(this.props.address);
+    this.setState({ mojiView: parseDna(moji.dna).view });
 
-        if (moji.isSire) {
-          this.setState({ isSire: true, isLoaded: true });
-          // if we know moji is a sire, break out of the promise chain
-          return Promise.reject('This moji is a sire');
-        } else {
-          return getSires(moji.owner);
```

Comment:
Probably a simpler way of handling this would have been to use a nested `then` chained off of this getSires rather than the wrapping `then` with this `catch` hack. Async/await works too though.

Reply:
Good to know! I was super unsure about how to handle it.. I did think of the nested Promises idea, but I fell too deeply into the mindset of "Promises are designed to prevent the pyramid of doom" and tried to look for any other solution. Looking back, I'd def say that nested Promises are significantly easier to understand than this wonky "reject out of the promise chain early" business.

It makes me wonder though, is there anything in other Promise implementations (eg, Bluebird) that allows for the termination of "the rest of the Promise chain"? 🤔

Reply:
I haven't heard of such a thing, but I wouldn't rule it out. I don't know Bluebird very well at all now that Promises are standard. Another use case I run into sometimes is "run this after the rest of the promise chain". Mithril.js does this with its redraw method. There is a built in HTTP library which returns a Promise. So you might call:
```
m.route.get('/some/route')
  .then(storeSomeData)
  .then(updateSomeValues)
  .catch(alertUser);
```
You never call `m.redraw`, but at the conclusion of that Promise chain Mithril automatically calls it, because they built this custom recursive structure that runs a callback after your last `then`. It's super clever, and feels broadly useful.